cairo-contracts/protobuf/src/types/message.cairo

```diff
@@ -0,0 +1,140 @@
+use protobuf::types::tag::{WireType, ProtobufTag, ProtobufTagImpl};
+use protobuf::primitives::numeric::NumberAsProtoMessage;
+use protobuf::types::wkt::Any;
+
+pub trait ProtoMessage<T> {
+    fn decode_raw(ref value: T, serialized: @ByteArray, ref index: usize, length: usize);
+    fn encode_raw(self: @T, ref output: ByteArray);
+    fn wire_type() -> WireType;
+    fn type_url() -> ByteArray;
+}
+
+pub struct ProtoCodec {}
+
+#[generate_trait]
+pub impl ProtoCodecImpl of ProtoCodecTrait {
+    fn decode<T, +ProtoMessage<T>, +Drop<T>, +Default<T>>(serialized: @ByteArray) -> T {
+        let mut index = 0;
+        let length = if ProtoMessage::<T>::wire_type() == WireType::LengthDelimited {
+            serialized.len()
+        } else {
+            0
+        };
+        let mut value = Default::<T>::default();
+        ProtoMessage::<T>::decode_raw(ref value, serialized, ref index, length);
+        value
+    }
+
+    fn encode<T, +ProtoMessage<T>>(value: @T) -> ByteArray {
+        let mut bytes = "";
+        value.encode_raw(ref bytes);
+        bytes
+    }
+
+    fn from_any<T, +ProtoMessage<T>, +Drop<T>, +Default<T>>(any: @Any) -> T {
+        if any.type_url != @ProtoMessage::<T>::type_url() {
+            panic!("unexpected type URL");
+        }
+        Self::decode::<T>(any.value)
+    }
+
+    fn to_any<T, +ProtoMessage<T>>(value: @T) -> Any {
+        Any { type_url: ProtoMessage::<T>::type_url(), value: Self::encode(value) }
+    }
+
+
+    fn decode_field<T, +ProtoMessage<T>, +Drop<T>>(
+        field_number: u8, ref value: T, serialized: @ByteArray, ref index: usize, bound: usize
+    ) {
+        assert(bound <= serialized.len(), 'invalid bound');
+        if index < serialized.len() && index < bound {
+            let tag = ProtobufTagImpl::decode(serialized[index]);
+            if tag.field_number == field_number {
```

Comment:
I think the protobuf specs allows fields to be encoded in any order, but here we assume that the fields are in ascending order. Also, Protobuf allows the same field tag to be present multiple times, by which the last occurance would win out.

In any case, I suppose the current implementation is sufficient for the first iteration for us to move on. We can figure a more robust decoding implementation later on.

For reference, Prost runs the decoding in a loop, with each field decoder repeatedly called when the tag matches. On the other hand, in Hermes SDK I tried a more organized approach, by first splitting the decode buffer into chunks called `ProtoChunk`, and then put them in a `BTreeMap` with the field tag being the key.